Add Concept Exercise: Windowing System (Classes) (#1476)
exercism/javascript: concepts/classes/about.md

JavaScript on the other hand did not have classes at all before they were added to the language specification in 2015. And even though a `class` keyword is available nowadays, JavaScript is still a _prototype-based_ language.

To understand what that means and how JavaScript actually works, we will go back to the time without classes.

## Prototype Syntax

In OOP, you want to create objects from "templates" so that they include certain data and functionality. The data (keys in the object) are called _fields_ in the OOP context, the functionality (functions in the object) are called _methods_.

### Constructor Function

In JavaScript, the template is facilitated by a regular function. Used in this context, it is called a _constructor function_ and the function name should start with a capital letter. Instances (objects) are derived from the template using the `new` keyword when invoking the constructor function.

```javascript
function Car() {
  // ...
}

const myCar = new Car();
const yourCar = new Car();
```

It is important to note that in JavaScript, the instances and the constructor function keep a relationship to each other even after the instances were created.

> Every object instance includes a hidden, internal object referred to as `[[prototype]]` in the language specification. It holds a reference to the value of the `prototype` key of the constructor function. Yes, you read that correctly, a JavaScript function can have key/value pairs because it is also an object behind the scenes. Nowadays `[[prototype]]` can be accessed via the key `__proto__` in many environments. It is important to not confuse the prototype of an instance (`[[prototype]]`/`__proto__`) with the `prototype` property of the constructor function.

Reviewer: do you prefer `__proto__` over `getPrototypeOf`

### Instance Fields

Often, you want all the derived objects to include some fields and pass some initial values when the object is constructed. This is facilitated via the `this` keyword. Inside the constructor function, `this` references the new object that will be created via `new` and it is automatically returned from the function.

That means we can add fields to the new instance by adding them to `this` in the constructor function.

```javascript
function Car(color, weight) {
  this.color = color;
  this.weight = weight;
  this.engineRunning = false;
}

const myCar = new Car('red', '2mt');
myCar.color;
// => 'red'
myCar.engineRunning;
// => false
```

> The syntax above the usual way of adding fields. In theory, instance fields that do not depend on the parameters of the constructor can also be added via the `prototype` entry of the constructor function.

Reviewer: add *is* between *above* and *the* - *above is the*

# About

JavaScript includes the capabilities for object-oriented programming ([OOP][wiki-oop]).
In OOP, you want to create objects (_instances_) from "templates" (_classes_) so that they include certain data and functionality.
The data properties are called _fields_ in the OOP context, the function properties are called _methods_.

> JavaScript did not have classes at all before they were added to the language specification in 2015.

Reviewer 2: ```suggestion
JavaScript did not have classes at all before they were added to the language specification in 2015, but allowed for object-oriented programming using prototype-based inheritance. 
```

Perhaps we clarify why this is mentioned at all.

Author: Good point!

> And even though a `class` keyword is available nowadays, JavaScript is still a _prototype-based_ language.

Reviewer 2: ...then this sentence makes more sense.

> To understand what that means and how JavaScript actually works, we will go back to the time when there were no classes.

Reviewer 2: ```suggestion
To understand what it means to be a prototype-based language and how JavaScript actually works, we will go back to the time when there were no classes.
```

I recommend we don't refer back to the previous sentences because the _that_ is slightly ambiguous.

## Prototype Syntax

### Constructor Function

In JavaScript, the template (class) is facilitated by a regular function.
When a function is supposed to be used as such a template, it is called a _constructor function_ and the convention is that the function name should start with a capital letter.
Instances (objects) are derived from the template using the `new` keyword when invoking the constructor function.

```javascript
function Car() {
  // ...
}

const myCar = new Car();
const yourCar = new Car();
```

It is important to note that in JavaScript, the instances and the constructor function keep a relationship to each other even after the instances were created.
Every instance object includes a hidden, internal property referred to as `[[prototype]]` in the language specification.
It holds a reference to the value of the `prototype` key of the constructor function.
Yes, you read that correctly, a JavaScript function can have key/value pairs because it is also an object behind the scenes.

Since 2015, `[[prototype]]` can be accessed via `Object.getPrototypeOf()`.
Before that, it was accessible via the key `__proto__` in many environments.

> Do not confuse the prototype of an object (`[[prototype]]`) with the `prototype` property of the constructor function.

Reviewer 2: Perhaps we recap?

```markdown
To summarize:

- Constructors in JavaScript are regular functions
- Constructing a new instance creates an object with a relation to its constructor called the prototype.
- Functions are objects (callable objects).
- Because functions are objects, it can have properties.
- The constructor's (function) prototype property will become the instance's prototype
```

Something like this? (Please check if correct).

Overall I like the `about` a lot, but I think it will help summarize the sections because it's a lot of "new" concepts.

Author: Great idea with the summary, I like it!

### Instance Fields

> Often, you want all the derived objects to include some fields and pass some initial values for those when the object is constructed.

Reviewer 2: ```suggestion
Often, you want all the derived objects (instances) to include some fields and pass some initial values for those when the object is constructed.
```

> This is facilitated via the [`this` keyword][mdn-this].

Reviewer 2: ```suggestion
This can be facilitated via the [`this` keyword][mdn-this].
```

Because the more advanced and flexible way is using `defineProperty` / `defineProperties`.

Inside the constructor function, `this` represents the new object that will be created via `new`.
`this` is automatically returned from the constructor function when it is called with `new`.

That means we can add fields to the new instance by adding them to `this` in the constructor function.

```javascript
function Car(color, weight) {
  this.color = color;
  this.weight = weight;
  this.engineRunning = false;
}

const myCar = new Car('red', '2mt');
myCar.color;
// => 'red'
myCar.engineRunning;
// => false
```

> ### Instance Methods

Reviewer 2: I think this is a very clear section. Nice.

Methods are added via the `prototype` property of the constructor function.
Inside a method, you can access the fields of the instance via `this`.
This works because of the following general rule.

> When a function is called as a method of an object, its `this` is set to the object the method is called on. [^1]

```javascript
function Car() {
  this.engineRunning = false;
  // ...
}

Car.prototype.startEngine = function () {
  this.engineRunning = true;
};

Car.prototype.addGas = function (litre) {
  // ...
};

const myCar = new Car();
myCar.startEngine();
myCar.engineRunning;
// => true
```

### The Prototype Chain

> `myCar` in the example above is a regular JavaScript object and if we would inspect it (e.g. in the browser console) we would not find a property `startEngine` with a function as value directly inside the `myCar` object.

Reviewer 2: ```suggestion
`myCar` in the example above is a regular JavaScript object and if we would inspect it (e.g. in the browser console), we would not find a property `startEngine` with a function as value directly inside the `myCar` object.
```

So how does the code above even work then?

The secret here is called the _prototype chain_.
When you try to access any property (field or method) of an object, JavaScript first checks whether the respective key can be found directly in the object itself.
If not, it continues to look for the key in the object referenced by the `[[prototype]]` property of the original object.
As mentioned before, in our example `[[prototype]]` points to the `prototype` property of the constructor function.

> That is where JavaScript would find the `startEngine` function because we added it there.

Reviewer 2: I think it makes sense to add the example in code again after this paragraph, to illustrate it.

And the chain does not end there.
The `[[prototype]]` property of `Car.prototype` (`myCar.[[prototype]].[[prototype]]`) references `Object.prototype` (the `prototype` property of the `Object` constructor function).
It contains general methods that are available for all JavaScript objects, e.g. `toString()`.

> In conclusion, you can call `myCar.toString()` and that method will exist because JavaScript searches for that method throughout the whole prototype chain.

Reviewer 2: Perhaps adding here that the prototype of `Object` is normally `null` which ends the chain?

You can find a detailed example in the [MDN article "Inheritance and the prototype chain"][mdn-prototype-chain-example].

> Note that the prototype chain is only travelled when retrieving a value.

Reviewer 2: Admonition here might be nice? To make it stand out.

Author: :+1: I used "caution".

> Setting or deleting a property of an instance object only targets that specific instance.

Reviewer 2: "Unlike most class-based inheritance" or something similar? I think we should point out the differences to what people are used with.

Perhaps also changes to _setting a property directly or deleting a property_, because you _can_ set properties via the prototype chain, but it's convoluted (requires a method that sets it, effectively).

Author: Added the first part you suggested as a separate sentence as it was hard to squeeze in.

> ### Dynamic Methods

Reviewer 2: ```suggestion
### Dynamic Methods
```

I don't know if this title is clear. Maybe it could be `Dynamic Methods / How to add methods to already existing instances, via their template`. I hate it, but that's the information I'm missing. Something explicit might also be written inside the paragraph below.

Author: You are right, it is not really clear what this is about. I adjusted the header and added a general sentence below.

Since 2015, `[[prototype]]` can be accessed via `Object.getPrototypeOf()`.
Before that, it was accessible via the key `__proto__` in many environments.

Do not confuse the prototype of an object (`[[prototype]]`) with the `prototype` property of the constructor function.

```exercism/note
To summarize:

- Constructors in JavaScript are regular functions.
- Constructing a new instance creates an object with a relation to its constructor called its _prototype_.
- Functions are objects (callable objects) and therefore they can have properties.
- The constructor's (function) `prototype` property will become the instance's _prototype_.
```

### Instance Fields

Often, you want all the derived objects (instances) to include some fields and pass some initial values for those when the object is constructed.
This can be facilitated via the [`this` keyword][mdn-this].
Inside the constructor function, `this` represents the new object that will be created via `new`.
`this` is automatically returned from the constructor function when it is called with `new`.

That means we can add fields to the new instance by adding them to `this` in the constructor function.

```javascript

> function Car(color, weight) {

Reviewer: Not sure if it is helpful to show how this is constructed under the hood as an object:

```
function Car(color, weight) {
    let thisObj = Object.create(null);
    thisObj.color = color;
    thisObj.weight = weight;
    return thisObj;
}
```
`this` removes the need to explicitly set the object and return it. Might be overkill or not needed :-)

Author: Since the student might not know about `Object.create` at this point, I felt this code is more confusing then helping. Imo the important part is that you don't have to explicitly return `this` and that is covered in the text above.

> `this` is automatically returned from the constructor function ...

Reviewer: Sounds great @junedev. This concept looks awesome and enjoyed reading the code when reviewing.

Reviewer 2: (fun fact, returning a _new_ object from the constructor function is in fact a way to get closures/private variables _and_ private functions from a constructor, however, you need to set the prototype at `Object.create` if you want to keep inheritance to work)
